doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,143 @@
+
+- initially after `kdbOpen`
+- after conflict errors in `kdbSet`
+
+While researching ways to do change tracking, we discovered that some sequences are currently legal but might be problematic.
+This issue not only effects change tracking, but may cause problems for other use cases too.
+
+Currently, plugins may (and in the case of notification plugins actually do) assume that a `kdbSet` operation for a certain parent key always directly follows the `kdbGet` operation for that same key. 
+However, this is not enforced in any way within Elektra, nor is it documented that it should.
+
+So in reality, as illustrated by the example, `kdbGet` and `kdbSet` of different parent keys might be interwoven.
+
+This is problematic, as plugins may store the result of `kdbGet` and use the stored data in the following `kdbSet` to calculate a changeset.
+Plugins have to do this, as there isn't any other Elektra-provided mechanism currently to do change-tracking for them.
+
+This problem is illustrated by the following example:
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the data below the parent `keyA`.
+- At point `(B)` the same thing happens for the data below the parent `keyB`.
+  This **replaces** the data from `(A)`.
+- At point `(C)` plugins that do change-tracking need to generate a changeset.
+  The changeset will be calculated between the stored data and the new data in the KeySet `a`.
+  Because the stored data is now from `(B)` and **not** from `(A)`, this changeset will be wrong.
+  The calculation will wrongly assume everything in KeySet `b` was removed and everything in KeySet `a` is newly added.
+
+
+To further illustrate the problem, assume KDB contains the following keys:
+
+```
+user:/a/brightness
+user:/a/saturation
+user:/b/background
+user:/b/foreground
+```
+
+`keyA` will be `user:/a`, `keyB` will be `user:/b`.
+Right after `(B)`, the two keysets are filled with the following values:
+
+| `a`                  | `b`                  |
+|----------------------|----------------------|
+| `user:/a/brightness` | `user:/b/background` |
+| `user:/a/saturation` | `user:/b/foreground` |
+
+For a plugin that does change-tracking as described above, the plugin will then wrongly calculate the following changeset at `(C)`:
+
+| Added Keys           | Removed Keys         | Modified Keys |
+|----------------------|----------------------|---------------|
+| `user:/a/brightness` | `user:/b/background` |               |
+| `user:/a/saturation` | `user:/b/foreground` |               |
+
+This is obviously wrong. Looking at the example, the changeset should be empty, as nothing has been changed in keyset `a`.
```

Comment:
Did you actually test this, e.g. with:

```
kdb mount a.ecf user:/a logchange
kdb mount b.ecf user:/b logchange
kdb set user:/a/brightness 1
kdb set user:/a/saturation 1
kdb set user:/b/background 1
kdb set user:/b/foreground 1
```

Or do you mean with other change tracking? Please be more specific and say what you actually checked.

Reply:
> e.g. with: [...]

This sequence would NOT be the sequence given above. Each `kdb set` is a `kdbGet` immediately followed by a `kdbSet` with the same KeySet and `parentKey`. That is the one case that always works without issues.

However, I can tell we don't need to test it. This is just a logical conclusion from what is stated above.

1. In line 56-59 we defined the contents of `a` and `b`
2. The last `kdbGet` at (B) used `b`, so that is "theirs"
3. The the current `kdbSet` at (C) uses `a`, so that is "ours"
4. Since "theirs" and "ours" are distinct, the changeset is just:
    - remove everything from "theirs"
    - add everything from "ours"

That's obviously wrong, because "theirs" is simply the wrong data. It should be the contents of `a` at (A), which would give the correct empty changeset.


All this follows from one single fact: Plugins store data internally that is entirely replaced during each `kdbGet` and then compared against during `kdbSet`.

Reply:
@markus2330 I have added a reproducible example